### PR TITLE
More restart policy options and consolidate batch/service restart tracker

### PR DESCRIPTION
**Pull request by dadgar**
More restart policy options and consolidate batch/service restart tracker

### CODE REVIEW

File: client/alloc_runner.go [hunk 1]
```diff
@@ -110,7 +110,7 @@ func (r *AllocRunner) RestoreState() error {
 		r.restored[name] = struct{}{}
 
 		task := &structs.Task{Name: name}
-		restartTracker := newRestartTracker(r.alloc.Job.Type, r.RestartPolicy)
+		restartTracker := NewRestartTracker(r.RestartPolicy)
```
Comment (diptanu) [inline client/alloc_runner.go]: should client depend on code from the api package?

Comment (dadgar) [inline client/alloc_runner.go]: This isn't from the api package. It is defined in the client package

File: client/restarts.go [hunk 2]
```diff
@@ -6,84 +6,53 @@ import (
...
-			startTime:   time.Now(),
-			interval:    restartPolicy.Interval,
-			delay:       restartPolicy.Delay,
-		}
-	default:
-		return &batchRestartTracker{
-			maxAttempts: restartPolicy.Attempts,
-			delay:       restartPolicy.Delay,
-		}
+func NewRestartTracker(policy *structs.RestartPolicy) *RestartTracker {
```
Comment (diptanu) [inline client/restarts.go]: Does this need to be exportable?

Comment (dadgar) [inline client/restarts.go]: I guess not. I can make it private

File: client/restarts.go [hunk 3]
```diff
@@ -6,84 +6,53 @@ import (
...
+	// Apply an artifical wait to enter the next interval
+	return r.shouldRestart(exitCode), end.Sub(now)
 }
 
-func (s *serviceRestartTracker) increment() {
-	s.count += 1
+// shouldRestart returns whether a restart should occur based on the exit code
+// and the RestartOnSuccess configuration.
+func (r *RestartTracker) shouldRestart(exitCode int) bool {
+	return exitCode != 0 || r.policy.RestartOnSuccess && exitCode == 0
```
Comment (diptanu) [inline client/restarts.go]: Do we even need to check if exitCode is 0 if restart on success is set?

**Pull request merged**